### PR TITLE
Recompute existing branches v2

### PR DESCRIPTION
**Pull request by lenzip**
This should enable overwriting branches with the same name, as per issue #81

### CODE REVIEW

**Comment by arizzi** [comment]
would this allow to read the value in the old tree?

**Comment by arizzi** [comment]
ah this is just the output branch, so I guess it works...

File: python/postprocessing/framework/output.py [hunk 1]
```diff
@@ -44,7 +45,7 @@ def __init__(self, tfile, ttree, intree):
         self._intree = intree
         self._branches = {} 
     def branch(self, name, rootBranchType, n=1, lenVar=None, title=None):
-        if (lenVar != None) and (lenVar not in self._branches) and (not self._tree.GetBranch(lenVar)):
+        if (lenVar != None) and (lenVar not in self._branches): #and (not self._tree.GetBranch(lenVar)):
```
Comment (arizzi) [inline python/postprocessing/framework/output.py]: @lenzip @gpetruc  why the last AND was needed / not needed anymore ?

**Comment by lenzip** [comment]
Hi Andrea
     the last "and" was making it impossible to overwrite the length var.
I'm unsure about the original need.
Giulio

**Comment by arizzi** [comment]
and why do we want to overwrite the len var? when we rewrite it, are we
sure that  all branches using that lenvar are of the correct length?

**Comment by lenzip** [comment]
For example, if we want to remove the leptons from the jets collection, 
the length of the jet collection changes.
In such a case one needs to make sure that all Jet_* branches are 
recomputed.
One example is here:
declaration of the branches: 
https://github.com/latinos/LatinoAnalysis/blob/master/NanoGardener/python/modules/LeptonSelLazy.py#L96-L98
writing of the branches: 
https://github.com/latinos/LatinoAnalysis/blob/master/NanoGardener/python/modules/LeptonSelLazy.py#L327-L334

**Comment by arizzi** [comment]
ok, but can we enforce that all collection are updated ?

**Comment by arizzi** [comment]
btw this also breaks the indices, so I think pruning collections (with same
name) should not be supported...

**Comment by lenzip** [comment]
On 08/05/2018 10:43, arizzi wrote:
> btw this also breaks the indices, so I think pruning collections (with 
> same
> name) should not be supported...
>
I'm afraid you are correct. I had not thought of this.
Should maybe some king of light of light cloning (e.g. just store the 
indices of the original objects in the pruned collection) be implemented?

**Comment by arizzi** [comment]
i.e. producing something like
selJetIdxes[newLen]

then accessing
Events->Draw("Jet_pt[selJetIdxes]")

**Comment by xjanssen** [comment]
Well we are already doing a custom made Lepton_XXX with XXX = pt, eta, phi , charge, type , iDx to Muon_ or Electron_ -> so we can avoid cleaning the original collections to be safe and we will do the same for Jets_ --> CleanJets_, still we will need to overwrite the   CleanJets_ and Lepton_ collections at some points but these would be our own collections, not the miniAOD ones.

**Comment by arizzi** [comment]
we could probably force some checks such that if you change the length then we automatically
- drop from output all the elements of the original collection
- drop from output all indices pointing there (the naming schema should allow to find that)

**Pull request merged**